sim/src/cap.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{CarID, VehicleType};
+use geom::{Duration, Time};
+use map_model::{LaneID, Map, Path, PathConstraints, PathStep};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+
+// Note this only indexes into the zones we track here, not all of them in the map.
+type ZoneIdx = usize;
+
+// This only caps driving trips.
+#[derive(Serialize, Deserialize, PartialEq, Clone)]
+pub struct CapSimState {
+    lane_to_zone: BTreeMap<LaneID, ZoneIdx>,
+    zones: Vec<Zone>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone)]
+struct Zone {
+    cap: usize,
+    entered_in_last_hour: BTreeSet<CarID>,
+    // TODO Maybe want sliding windows or something else
+    hour_started: Time,
+}
+
+impl CapSimState {
+    pub fn new(map: &Map) -> CapSimState {
+        let mut sim = CapSimState {
+            lane_to_zone: BTreeMap::new(),
+            zones: Vec::new(),
+        };
+        for z in map.all_zones() {
+            if let Some(cap) = z.restrictions.cap_vehicles_per_hour {
+                let idx = sim.zones.len();
+                for r in &z.members {
+                    for l in map.get_r(*r).all_lanes() {
+                        if PathConstraints::Car.can_use(map.get_l(l), map) {
+                            sim.lane_to_zone.insert(l, idx);
+                        }
+                    }
+                }
+                sim.zones.push(Zone {
+                    cap,
+                    entered_in_last_hour: BTreeSet::new(),
+                    hour_started: Time::START_OF_DAY,
+                });
+            }
+        }
+        sim
+    }
+
+    pub fn car_entering_lane(&mut self, now: Time, car: CarID, lane: LaneID) {
```

Comment:
I started to add a handler for when cars leave the zone, but I dunno if that's useful yet. Maybe we'll want to express a policy like "only max 1000 cars in entire downtown area at a time" later.